### ISSUE TITLE
Printing recursive types

### ISSUE THREAD

**Issue by dariusf** [whitequark/ppx_deriving#61]
Is there a way to generate printers for recursive types?

For example,
```ocaml
type 'a exp = [
  | `Plus of 'a * 'a
  | `Num of int
] [@@deriving show]

let rec eval = function
  | `Plus (e1, e2) -> eval e1 + eval e2
  | `Num i -> i

;;

let ast = (`Plus (`Plus (`Num 1, `Num 2), `Plus (`Num 3, `Num 4))) in
let result = eval ast in
print_endline @@ show_magic @@ ast;
print_endline @@ string_of_int @@ result;
```
I'm guessing the printer would look something like `eval`, but would compose a given `pp` function. Don't understand `ppx_deriving` well enough to know how difficult this is though.

**Comment by whitequark** [comment]
Yes:
```
# let ast = (`Plus (`Plus (`Num 1, `Num 2), `Plus (`Num 3, `Num 4)));;
val ast : [> `Plus of [> `Plus of [> `Num of int ] * [> `Num of int ] ] * [> `Plus of [> `Num of int ] * [> `Num of int ] ] ] =
  `Plus (`Plus (`Num 1, `Num 2), `Plus (`Num 3, `Num 4))
# type exprec = [
  | `Plus of exprec * exprec
  | `Num of int
] [@@deriving show];;
# show_exprec ast;;
- : bytes = "`Plus ((`Plus ((`Num (1), `Num (2))), `Plus ((`Num (3), `Num (4)))))"  
```

It would be nice if you could do something like `[%show: [ `Plus of 'a * 'a | `Num of int ] as 'a]` without introducing a type, but currently that doesn't work. In fact it breaks in a fairly opaque way:
```
# [%show: [ `Plus of 'a * 'a | `Num of int ] as 'a];;
Error: Unbound value poly_a        
```
which is something I should fix.

**Comment by dariusf** [comment]
Ah, I see. I'm defining it in this way because I want to tie the knot later, in order to extend the variant.

```ocaml
type 'a exp1 = [
  | 'a exp
  | `Times of 'a * 'a
]
```
If it's too niche a use case then I'll write the printer by hand.

Would the `as 'a` example desugar to a named type, like `exprec`?

**Comment by whitequark** [comment]
Well, I'm suggesting to define the printer at the time you're tying the knot. The printer I wrote is no more or less polymorphic than your eval function (which also ties the knot), no?

Try writing an eval function that *wouldn't* tie the knot to see what I mean; it's similarly impossible to write a show function like that.

As for the inline form, I've no idea how to actually implement it, as I did not think about it yet.
